### PR TITLE
Unify session receipt types into a single SessionReceipt type

### PR DESCRIPTION
**Pull request by SchmErik**
This done by abstracting over segment receipts via the new `SegmentReceipt`
trait and using the same `verify` function for both receipt types. Now each
`SessionReceipt` contains a vector of `SegmentReceipt` trait objects. This
simplifies the Receipt types to facailitate dev-mode and dogfooding.

This PR switches the default hash function to be poseidon (which is more friendly for recursion).
In the future we need to clean up the verifier so that we can more easily allow usage of different hash functions.

### CODE REVIEW

File: risc0/zkvm/src/prove/mod.rs [hunk 1]
```diff
@@ -292,52 +292,53 @@ where
...
-        table.insert("cpu".to_string(), prover.clone());
-        table.insert("$default".to_string(), prover);
+        let prover = Rc::new(LocalProver::new("cpu:sha256", cpu::sha256_hal_eval()));
+        table.insert("cpu:sha256".to_string(), prover.clone());
+        table.insert("$sha256".to_string(), prover);
 
         let prover = Rc::new(LocalProver::new("cpu:poseidon", cpu::poseidon_hal_eval()));
         table.insert("cpu:poseidon".to_string(), prover.clone());
-        table.insert("$poseidon".to_string(), prover);
+        table.insert("$poseidon".to_string(), prover.clone());
```
Comment (mothran) [inline risc0/zkvm/src/prove/mod.rs:315]: warning: we found poseidon to be far slower on CPU than the sha256 hashing so this change will probably slow down the majority of local proving.

Comment (SchmErik) [reply risc0/zkvm/src/prove/mod.rs:315]: Bummer! Thanks for the heads up. I saw the perf numbers and was thinking that it was something else..

Comment (SchmErik) [reply risc0/zkvm/src/prove/mod.rs:315]: I'll make a quick fix and bring this up internally

Comment (flaub) [reply risc0/zkvm/src/prove/mod.rs:315]: The related benchmark shows a significant improvement in performance, we might be reading the results wrong

**Comment by flaub** [comment]
Superseded by #666

**Pull request closed without merging**